rs/linux_kernel_command_line/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+            self.val
+        )
+    }
+}
+
+#[derive(Debug)]
+/// A value unrepresentable as a kernel command line argument value.
+pub struct UnrepresentableValue {
+    val: String,
+}
+
+impl StdError for UnrepresentableValue {}
+
+impl fmt::Display for UnrepresentableValue {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Illegal characters in supplied value {:?}", self.val)
+    }
+}
+
+#[derive(Debug, Default)]
+/// Represents a correctly-parsed kernel command line.
+pub struct KernelCommandLine {
+    tokenized_arguments: Vec<String>,
+}
+
+impl KernelCommandLine {
+    fn format_argument(
+        argument: &str,
+        value: Option<&str>,
+    ) -> Result<String, UnrepresentableValue> {
+        Ok(argument.to_string()
+            + &match value {
+                Some(val) => ("=".to_owned()
+                    + (if val.contains("\"") || val.contains("\n") {
+                        return Err(UnrepresentableValue {
+                            val: val.to_string(),
+                        });
+                    } else if val.contains(" ") {
+                        ("\"".to_owned() + val + "\"").to_string()
+                    } else {
+                        val.to_string()
+                    }
+                    .as_str()))
+                .to_string(),
+                None => "".to_string(),
+            })
+    }
+
+    /// Remove an argument from a kernel command line, however many times it appears.
+    /// Returns the position of the first removed argument.
+    pub fn remove_argument(&mut self, argument: &str) -> Option<usize> {
+        let mut firstpos: Option<usize> = None;
+        self.tokenized_arguments = self
+            .tokenized_arguments
+            .iter()
+            .enumerate()
+            .filter(|(pos, arg)| {
+                let res =
+                    *arg != argument && !arg.starts_with((argument.to_owned() + "=").as_str());
+                if !res && firstpos.is_none() {
+                    firstpos.replace(*pos);
+                }
+                res
+            })
+            .map(|(_, x)| x.clone())
+            .collect::<Vec<String>>();
+        firstpos
+    }
+
+    /// Ensure an argument is present in a kernel command line.
+    /// If the argument is in the command line, it is replaced with the supplied
+    /// value in the same position, and all other mentions are removed.
+    /// If the argument is not in the command line, it is appended to the end.
+    pub fn ensure_single_argument(
+        &mut self,
+        argument: &str,
+        value: Option<&str>,
+    ) -> Result<(), UnrepresentableValue> {
+        match self.remove_argument(argument) {
+            Some(pos) => {
+                // match pos >= self.tokenized_arguments.len()
+                self.tokenized_arguments
+                    .insert(pos, Self::format_argument(argument, value)?);
```

Comment:
Perhaps doesn't matter a lot, but if `format_argument` returns an error, `self` won't have an arguments with name `argument` (the existing args are removed on line 95 but we can't add the new arg). If this is an issue, you could extract `Self::format_argument(argument, value)?` into a variable and move it before the call to `remove_argument`.

Reply:
Not sure I understood but I extracted the format argument thing to a separate variable.

![image](https://github.com/user-attachments/assets/1a0cba9f-0467-40d6-b5a1-bf9d747e5c54)